### PR TITLE
Creating a multi-task learning example for ranking models

### LINKED ISSUE #687: [Task] Create Multi-Task Learning example
### Description
Create an example notebook on how to use Multi-Task Learning (MTL) with Merlin Models.

**Additional context**
- Models currently provides support to `MMOEBlock`, `CGCBlock`, `PLEBlock`, which creates specific and shared experts for MTL tasks.
- This example will be based in the scripts created in our MTL recsys research.
- Will use TenRec synthetic data

### PR DESCRIPTION
**Pull request by gabrielspmoreira**
Fixes #687 

### Goals :soccer:
This PR creates a multi-task learning (MTL) example for Merlin Models.

### Implementation Details :construction:
- The example is based on the `ModelOutput` based API for prediction tasks, in particular in the new `OutputBlock` introduced in #772 .
- The example introduces many building blocks related to MTL, like `ColumnBasedSampleWeight`, the options to set task-specific towers, sample weights and class weights, the ability to condition the loss of one task on the sample space from another task, and the MTL-specific architectures introduced in #902 : MMoE, CGC and PLE.
- The example uses synthetic data based on the schema of [TenRec dataset](https://arxiv.org/pdf/2210.10629.pdf). We add support to `tenrec-video` to `generate_data()` function.

### Testing Details :mag:
- Created a unit test `test_usecase_ranking_with_multitask_learning` to check for notebook execution and final metrics.

### CODE REVIEW

File: examples/usecases/ranking_with_multitask_learning.ipynb [hunk 1]
```diff
@@ -0,0 +1,1663 @@
+{
```
Comment (bschifferer) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>Is the benefit to require less computational resources or are there additional benefits like higher accuracy in some targets?</p><p> </p><p>Can we add the real word (+reference) for MMoE, CGC and PLE? I do not know what the abbreviations are :)</p><p> </p><p>E.g. [Multi-gate Mixture-of-experts (MMOE)](link), etc.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (bschifferer) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>If I understand it correctly, this use case it to weight the loss of each sample given another column. </p><p> </p><p>I think the example explains the functionality, but using "age" or "watching_times" as an example is pretty confusing. I do not think we would use that in the real-world. I would rather change the example to provide a sample_weight column and provide a mapping of age to weights.</p><p> </p><p>E.g. </p><p>Age 0-17: 1.0, </p><p>Age: 18-25: 10.0, </p><p>Age: 25-35: 5.0, </p><p>Age: 35+ : 1.0</p><p> </p><p>With an example that we want to focus on our target group and weight their examples higher. I think that is easier to understand</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (bschifferer) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>I wonder if it makes sense (or it is possible) to describe the different use case with LaTeX in mathematical expression?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (bschifferer) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>Why do we combine CGC and PLE? I would expect another header PLE as it has its own API./ description and parameters</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

Comment (bschifferer) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>Can we add a summary paragraph and (next step paragraph) as in many other examples?</p><p> </p><p>Not sure if next steps are used in the use cases - as they are often single examples</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

**Review by bschifferer — CHANGES_REQUESTED** [review]
@radekosmulski can you help @gabrielspmoreira to provide a logo link for the example.

Currently, it uses https://developer.download.nvidia.com/notebooks/dlsw-notebooks/merlin_models_03-exploring-different-models/nvidia_logo.png which is incorrect

**Comment by bschifferer** [comment]
Thanks for the example. That is a great introduction to Multi-Task Learning

File: examples/usecases/ranking_with_multitask_learning.ipynb (hunk 1 above)
Comment (radekosmulski) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>minor typos: "to train a single model that is" (vs current "are") and then "the building blocks Merlin Models provides" (vs current "provide")</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (radekosmulski) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>typo: "there might <strong>be </strong>some event dependency in the system the user is interacting <strong>with</strong>"</p><p> </p><p>BTW this is a really cool feature!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (radekosmulski) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>"projecting the inputs independently and then having the experts output<s>s</s> weighted averaged by a gate <s>to</s> for a shared representation to be used for all tasks" (two typos, s and to are unnecessary)</p><p> </p><p>"gate for each task, so that they could choose how <strong>to </strong>better combine the experts outputs" ("to" missing)</p><p> </p><p>" dataset<s> by</s> with just a few lines of code!" </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Comment (radekosmulski) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>"the<strong>y</strong> proposed allowing for some"  ("y" missing)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

Comment (radekosmulski) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>"multiple <strong>CGC</strong> model<strong>s</strong>" ("s" missing)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>Sure. I have clarified those points now</p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>Thanks for you review @radekosmulski&nbsp;. Fixed those typos</p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>thanks @radekosmulski&nbsp;. We found these scenarios of some targets depending on other targets in our research experiments with AliCCP and TenRec. It seems a good practice, because without that the results for sparser targets are really poor</p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>If you are referring to explaining the different sample weights use cases with math, I think that would be trick.</p><p>The formula to compute sample weight itself is simple, it either use another column as sample weight, or convert a binary column to sample weight by doing <code>tf.where(binary_column_value == 1, pos_weight, neg_weight)</code>. Not sure if that helps.</p><p>The rest of the logic of <code><a href="https://github.com/NVIDIA-Merlin/models/blob/f0bb535046a6de008f182caede2fd06341a26267/merlin/models/tf/outputs/block.py#L231-L285" target="_blank">ColumnBasedSampleWeight</a></code> is to deal with missing arguments and extracting the values for <code>weight_column_name</code> from features/target. </p><p> </p><p> </p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>fixed</p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>fixed</p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>fixed</p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>The reason is that both CGC and PLE were introduced in the same paper. PLE is the multi-level version of CGC.</p><p>I have added this info to the beginning of that section:  "The&nbsp;<strong>CGC</strong>&nbsp;and&nbsp;<strong>PLE</strong>&nbsp;architectures were introduced in this&nbsp;<a href="https://dl.acm.org/doi/10.1145/3383313.3412236" target="_blank">p</a>aper&nbsp;(2020)."</p><p>Also added subsections for CGC and PLE</p><p>Does that help to clarify?</p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>Sure. I have added a conclusion</p>

Comment (gabrielspmoreira) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>Your are right. That example is confusing.  </p><p>I am using the schema for TenRec, I think creating another column and encoding it just for this sampling weight use case could be a detour.</p><p>I simplified this use case to a single task (below) and as watching_times is the count of times the user watches the video, it indicates the strength of the relevance of that item for the user, and it could be reasonable using that column as sample weight.</p><p> </p><pre class="ql-syntax" spellcheck="false">output_block = mm.BinaryOutput("click",
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; post=mm.ColumnBasedSampleWeight("watching_times")
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;)
</pre>

File: tests/unit/tf/examples/test_usecase_ranking_with_multitask_learning.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import pytest
+from testbook import testbook
+
+from tests.conftest import REPO_ROOT
+
+
```
[the remark this answers is not in the document]
Comment (gabrielspmoreira) [reply tests/unit/tf/examples/test_usecase_ranking_with_multitask_learning.py:6]: There is this line in `tests/unit/tf/__init__.py` which means that no unit tests are run if tensorflow is not installed
`pytest.importorskip("tensorflow")`

**Comment by gabrielspmoreira** [comment]
I have updated the notebook to use the following logo image URL, by following the convention of the existing URLs:
```
https://developer.download.nvidia.com/notebooks/dlsw-notebooks/merlin_models_ranking_with_multitask_learning/nvidia_logo.png
```

assuming that will be the URL to be created later.

File: examples/usecases/ranking_with_multitask_learning.ipynb (hunk 1 above)
Comment (bschifferer) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: I like that example, that make more sense

Comment (bschifferer) [reply examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>Yes, I like it with the subheading. I understand that you wrote subheading per paper. I think more from the user perspective - as an user, I am interested in the different architecture designs and then test them (all or 1-by-1) and might prioritize them - independent if they are from the same paper or different papers?</p>

File: examples/usecases/ranking_with_multitask_learning.ipynb [hunk 3]
```diff
@@ -0,0 +1,1707 @@
...
```
Comment (bschifferer) [inline examples/usecases/ranking_with_multitask_learning.ipynb:1]: <p>I would add this paragraph after the heading</p><p>So instead of having all tasks sharing all the experts, they proposed allowing for some task-specific experts and shared experts, which they named <strong>Customized Gate Control (CGC) Model</strong>, for which we provide a building block.    &nbsp;Notice that <code>CGCBlock</code> has separate arguments for <code>num_task_experts</code> and <code>num_shared_experts</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/948/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='50'/>

**Comment by bschifferer** [comment]
> I have updated the notebook to use the following logo image URL, by following the convention of the existing URLs:
> 
> ```
> https://developer.download.nvidia.com/notebooks/dlsw-notebooks/merlin_models_ranking_with_multitask_learning/nvidia_logo.png
> ```
> 
> assuming that will be the URL to be created later.

@radekosmulski can you create a logo?
@gabrielspmoreira I think the new url format does not use _ in the title anymore. Let's wait if Radek can provide a logo link tomorrow?

**Review by bschifferer — APPROVED** [review]

**Comment by radekosmulski** [comment]
hi @gabrielspmoreira! As request by @bschifferer, I created a tracking logo for the new notebook. Could you please replace the URL to the logo in the notebook description with the following:

`https://developer.download.nvidia.com/notebooks/dlsw-notebooks/merlinmodelsrankingwithmultitasklearning/nvidia_logo.png`

Thi is a correct tracking logo, for some reason the system started to encode the release names like this. But as you can see from the screenshot below, it maps to a properly created release on the CDN system:

image
